[#371] Database Engine Base App: Ability to run SQL queries, view results, and inspect database structure within database catalog

Na-daT opened the pull request:
## Description
Add generic SQL query reactor with permission-based validation and delegation to handle both SELECT and modification queries through a single entry point.

## Changes Made
- Created SqlQueryReactor file.
- Added the new reactor to the ReactorFactory.

## How to Test
Run the pixel command in the terminal app, or navigate to Database Catalogue > your database engine > Query tab.

**Sample Pixel commands:**

- **Select Query**
`SqlQuery(database=["database-engine-id"], query=["SELECT * FROM users"], limit=[100])`

- **Modification Queries**
`SqlQuery(database=["database-engine-id"], query=["INSERT INTO users (name) VALUES ('test')"], commit=[true])`

**Expected Results:**
 
- SELECT queries return data with limit
- Modification queries show true/false
- Permission errors for unauthorized operations

Na-daT committed:
[#371] build: created generic sql query reactor

CodiumAI-Agent commented:
## Title

[#371] Database Engine Base App: Ability to run SQL queries, view results, and inspect database structure within database catalog

___
### **User description**
## Description
Add generic SQL query reactor with permission-based validation and delegation to handle both SELECT and modification queries through a single entry point.

## Changes Made
- Created SqlQueryReactor file.
- Added the new reactor to the ReactorFactory.

## How to Test
Run the pixel command in the terminal app, or navigate to Database Catalogue > your database engine > Query tab.

**Sample Pixel commands:**

- **Select Query**
`SqlQuery(database=["database-engine-id"], query=["SELECT * FROM users"], limit=[100])`

- **Modification Queries**
`SqlQuery(database=["database-engine-id"], query=["INSERT INTO users (name) VALUES ('test')"], commit=[true])`

**Expected Results:**
 
- SELECT queries return data with limit
- Modification queries show true/false
- Permission errors for unauthorized operations


___

### **PR Type**
Enhancement


___

### **Description**
- Provide unified SQL query entrypoint (`SqlQuery` pixel)

- Parse SQL with JSQLParser, detect query type

- Validate user permissions per query type

- Delegate to select or modification reactors


___

### **Changes diagram**

```mermaid
flowchart LR
  A["Input SQL Query"] --> B["Detect QueryType"]
  B --> C["Validate Permissions"]
  C --> D["SELECT -> BasicIteratorTask"]
  C --> E["MODIFY -> ExecQueryReactor"]
```


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ReactorFactory.java</strong><dd><code>Register SqlQueryReactor in factory</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/prerna/reactor/ReactorFactory.java

<li>Imported <code>SqlQueryReactor</code> class<br> <li> Registered <code>"SQLQuery"</code> in reactorHash<br> <li> Updated reactor factory mappings


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/959/files#diff-f7abecd3a8ba806478205c7c6ac08e84c08d39b28fb242d186552830498062e4">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>SqlQueryReactor.java</strong><dd><code>Implement unified SQLQueryReactor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/prerna/reactor/qs/SqlQueryReactor.java

<li>Added <code>SqlQueryReactor</code> extending <code>AbstractReactor</code><br> <li> Implements SQL parsing via JSQLParser<br> <li> Validates user permissions for SELECT vs modifications<br> <li> Delegates to select or ExecQuery reactors


</details>


  </td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605">+230/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

CodiumAI-Agent commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R162-R162'><strong>Typo in method name</strong></a>

The method `addQueriedDatabasesese` appears misspelled and may not correctly track queried databases.
</summary>

```java
this.insight.addQueriedDatabasesese(databaseId);
```

</details>

<details><summary><a href='https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R206-R206'><strong>Default commit behavior</strong></a>

Commit logic defaults to true when `commitStr` is null or empty, potentially causing unintended commits.
</summary>

```java
boolean commit = commitStr == null || commitStr.trim().isEmpty() || Boolean.parseBoolean(commitStr);
```

</details>

<details><summary><a href='https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R137-R139'><strong>Unhandled SQL types</strong></a>

Statements classified as OTHER (e.g., DDL) are treated as modification queries, which may not execute correctly or as intended.
</summary>

```java
} else {
	return executeModificationQuery(sqlQuery, databaseId, commitStr);
}
```

</details>

</td></tr>
</table>

CodiumAI-Agent commented:
## PR Code Suggestions ✨

<!-- 568ef44 -->

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Fix typo in method name</summary>

___


**Correct the method call typo to match the actual Insight API. Likely the intended <br>method is <code>addQueriedDatabases</code>. This ensures compilation and proper tracking of <br>queried databases.**

[src/prerna/reactor/qs/SqlQueryReactor.java [162]](https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R162-R162)

```diff
-this.insight.addQueriedDatabasesese(databaseId);
+this.insight.addQueriedDatabases(databaseId);
```
<details><summary>Suggestion importance[1-10]: 10</summary>

__

Why: The typo in `addQueriedDatabasesese` prevents compilation, so correcting the method call is critical for functionality.


</details></details></td><td align=center>High

</td></tr><tr><td>



<details><summary>Handle unsupported SQL types explicitly</summary>

___


**Treat unsupported SQL types explicitly instead of routing all non-SELECT queries to <br>modification logic. This avoids executing DDL statements unexpectedly.**

[src/prerna/reactor/qs/SqlQueryReactor.java [132-139]](https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R132-R139)

```diff
-private NounMetadata delegateToAppropriateReactor(String sqlQuery, String databaseId, 
+private NounMetadata delegateToAppropriateReactor(String sqlQuery, String databaseId,
         QueryType queryType, String limitStr, String commitStr) {
-    
-    if (queryType == QueryType.SELECT) {
-        return executeSelectQuery(sqlQuery, databaseId, limitStr);
-    } else {
-        return executeModificationQuery(sqlQuery, databaseId, commitStr);
+    switch (queryType) {
+        case SELECT:
+            return executeSelectQuery(sqlQuery, databaseId, limitStr);
+        case INSERT:
+        case UPDATE:
+        case DELETE:
+            return executeModificationQuery(sqlQuery, databaseId, commitStr);
+        default:
+            throw new SemossPixelException("Unsupported SQL statement type: " + queryType);
     }
 }
```
<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: Explicitly rejecting `OTHER` query types prevents executing unsupported DDL as modifications, securing correctness.


</details></details></td><td align=center>Medium

</td></tr><tr><td rowspan=2>General</td>
<td>



<details><summary>Validate non-negative limit</summary>

___


**Add a check to ensure parsed limits are positive and reset to default if zero or <br>negative. This prevents unintended unlimited or invalid fetch sizes.**

[src/prerna/reactor/qs/SqlQueryReactor.java [151-157]](https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R151-R157)

```diff
 if (limitStr != null && !limitStr.trim().isEmpty()) {
     try {
         limit = Integer.parseInt(limitStr.trim());
+        if (limit <= 0) {
+            classLogger.warn("Non-positive limit value: " + limit + ", using default 500");
+            limit = 500;
+        }
     } catch (NumberFormatException e) {
         classLogger.warn("Invalid limit value: " + limitStr + ", using default 500");
     }
 }
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: Ensuring the `limit` is positive avoids unintended zero or negative fetch sizes, improving robustness.


</details></details></td><td align=center>Low

</td></tr><tr><td>



<details><summary>Replace magic string with enum key</summary>

___


**Use the defined enum key instead of a magic string for creating the commit noun. <br>This ensures consistency and avoids typos.**

[src/prerna/reactor/qs/SqlQueryReactor.java [208]](https://github.com/SEMOSS/Semoss/pull/959/files#diff-347d4ead6f5f5891c81aeb4c8c60184a42e6032575ed1167d0200a859b53e605R208-R208)

```diff
-GenRowStruct commitGrs = this.store.makeNoun("commit");
+GenRowStruct commitGrs = this.store.makeNoun(ReactorKeysEnum.COMMIT.getKey());
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: Using `ReactorKeysEnum.COMMIT.getKey()` instead of the literal `"commit"` improves consistency and reduces typo risk.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

ibrahimelnemr committed:
Merge branch 'dev' into 371-database-engine-run-sql

ibrahimelnemr committed:
Merge branch 'dev' into 371-database-engine-run-sql

Na-daT committed:
[#371] fix: defaulted commit to false and added validation checks in SqlQueryReactor

Na-daT committed:
[#371] fix: add task optimization for select queries in SqlQueryReactor

Na-daT committed:
Merge branch 'dev' into 371-database-engine-run-sql

Na-daT committed:
feat: get database category reactor (#371)

- limit sql query fetch to 50 rows
- create getDatabaseCategory to identify sql or noSQL db

Na-daT committed:
Merge remote-tracking branch 'origin/dev' into 371-database-engine-run-sql

ibrahimelnemr committed:
Merge branch 'dev' into 371-database-engine-run-sql

themaherkhalil committed:
Merge branch 'dev' into 371-database-engine-run-sql

themaherkhalil committed:
fix: inner reactors have their own independent noun stores, consolidating some methods

themaherkhalil committed:
chore: reverting back reactor factory changes
